### PR TITLE
Moment.js Timestamp

### LINKED ISSUE #1280: Timestamp: display time (when published > 10h && is today)
**Is your feature request related to a problem? Please describe.**
We currently don't output the time in our timestamps - only the date. But under certain conditions, we do want to display the time (in BST, or in GMT, depending on when the article was published). NB, we're happy showing BST time to .com users (i.e. non-UK users).

Rule should be: 
- if it's yesterday or prior, given the timezone, we show the date
- if it's today, given the timezone, we show the date and time 
- if it's today, given the timezone, and less than 10 hours ago, we show the relative time.

Pseudocode for displaying absolute vs relative timestamps:

```
if ( (timestamp is lastPublished) || ( (timestamp is firstPublished) && (there is no lastPublished) ) )
    relativeTimeIsPossible

if (relativeTimeIsPossible && timestamp < 10 hrs ago) 
    displayRelativeTime() // e.g. "4 minutes ago"

else if (timestamp is from today)
    displayAbsoluteDateTime() // e.g. "4 March 2019, 15:14 BST"

else
    displayAbsoluteDate() // e.g. "4 March 2019"
```

Please note that this issue is NOT about the prefix text ("Updated") - this is simply to do with the datetime portion of the timestamp.

**Describe the solution you'd like**

The code for displaying absolute and relative time has been done, but the code for switching between the two has not - that's where this issue comes in. In addition, we'll need to add a formatting method to handle the time display, and import a timezone library to handle conversion to BST/GMT.

So we should be able to display all of these:

- `4 minutes ago`
- `13 February 2019`
- `13 February 2019, 14:36 GMT`
- `13 April 2019, 14:36 BST`

**Describe alternatives you've considered**
N/A

**Testing notes**
Cypress:
- We want to test content of the text, making sure that it is displaying correctly based on the information pulled in from Ares.

Manual:
- Test across devices/browsers to make sure it displays as expected and has no clipping issues, etc.

Accessibility:
- Use AT to make sure that the text is read out as expected.
- Test Colour Ratio.

We want to test content of the text, but will decide location between unit/cypress in issue refinement

James wants to test the behaviour with both a minor & major update, to make sure lastPublished only changes on major changes.

Potential edge case of GMT/BST change. If published in, hour is always the BST value. “X hours ago…” should always be the “real” value, not accounting for the +/- hour.

Dev insight: Will need lots of unit tests.

**Additional context**
[As described by Catharine](https://github.com/bbc/simorgh/issues/1282#issuecomment-463385663), we should ensure that "BST" & "GMT" are read out suitably on accessibility devices.

Probably worth branching off https://github.com/bbc/simorgh/pull/1519.

- [x] Initially labelled with ["Refinement needed"](https://github.com/bbc/simorgh/labels/Refinement%20Needed)

### PR DESCRIPTION
**Pull request by twinlensreflex**
Resolves #1280

**Overall change:** _Adds moment and adds the time to the date when less than 24 hours ago._

**Code changes:**

- Moment has been added using the [webpack magic method](https://github.com/bbc/simorgh/issues/1280#issuecomment-478534463)
- I've added a test for `'should display date and time when updated > 10 hours ago && < 24 hours ago'`
- I've implemented the Moment logic for all current timestamp scenarios
- I've implemented Moment logic for the new scenario so that the test passes

Testing Notes:
These scenarios will need to be tested:
1. with an article published less than 10 hours ago timestamps should be `4 minutes ago`
2. with an article published before today timestamps should be `13 February 2019`
3. with an article published today (when today is GMT) but more than 10 hours ago timestamp format should be `13 February 2019, 14:36 GMT`
3. with an article published today (when today is BST) but more than 10 hours ago timestamp format should be `13 April 2019, 14:36 BST`
---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

File: src/app/containers/Timestamp/index.jsx [hunk 1]
```diff
@@ -5,16 +5,23 @@ import relativeTime from './relativeTimestamp';
...
+  alphaNumericDatetime,
   isValidDateTime,
   formatUnixTimestamp,
   isTenHoursAgoOrLess,
+  isTwentyFourHoursAgoOrLess,
 } from './timestampUtilities';
 import { GridItemConstrainedMedium } from '../../lib/styledGrid';
 
+const formatType = timestamp =>
+  isTwentyFourHoursAgoOrLess(timestamp)
```
Comment (twinlensreflex) [inline src/app/containers/Timestamp/index.jsx]: I tried to replace this with moment logic but it seems as if the 24 hour check is more accurate, both [today.diff(timestamp) <= 0](https://stackoverflow.com/questions/21284312/moment-js-check-if-a-date-is-today-or-in-the-future) and [today.isSame(timestamp, 'd')](https://stackoverflow.com/questions/21284312/moment-js-check-if-a-date-is-today-or-in-the-future) cause the related tests to fail.

File: package.json [hunk 2]
```diff
@@ -110,6 +110,9 @@
     "helmet": "^3.14.0",
     "isomorphic-fetch": "^2.2.1",
     "js-cookie": "^2.2.0",
+    "moment": "^2.24.0",
```
Comment (dr3) [inline package.json]: We shouldnt need `moment` as its inc in `moment-timezone` :)

**Review by dr3 — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (twinlensreflex) [reply package.json]: 👍 thanks

File: src/app/containers/Timestamp/helpers/testHelpers.js [hunk 3]
```diff
@@ -15,4 +18,4 @@ const timestampGenerator = timeDifference => {
   return timestamp;
 };
 
-export default timestampGenerator;
+export const isBritishSummerTime = timestamp => moment(timestamp).isDST();
```
Comment (dr3) [inline src/app/containers/Timestamp/helpers/testHelpers.js]: You might need to specify a region here. If this runs in the US, its idea of when DST is will be different

File: src/app/containers/Timestamp/index.test.jsx [hunk 4]
```diff
@@ -95,9 +85,12 @@ describe('Timestamp', () => {
     expect(renderedWrapper[1].children[0].data).toEqual('Updated 4 hours ago');
   });
 
-  it('should render absolute time for lastPublished and for firstPublished if > 10 hrs ago', () => {
-    const firstPublishedTwelveHoursAgo = timestampGenerator({ hours: 12 });
-    const lastPublishedElevenHoursAgo = timestampGenerator({ hours: 11 });
+  it('should render absolute time (without datetime) for lastPublished and for firstPublished if published today > 10 hrs ago', () => {
+    const firstPublishedTwelveHoursAgo = timestampGenerator({
```
Comment (dr3) [inline src/app/containers/Timestamp/index.test.jsx]: Surely this is 27 hours, not 12?

File: src/app/containers/Timestamp/index.test.jsx [hunk 5]
```diff
@@ -111,4 +104,60 @@ describe('Timestamp', () => {
       /Updated [0-9]{1,2} \w+ [0-9]{4}/,
     );
   });
+
+  describe('time dependent tests', () => {
```
Comment (dr3) [inline src/app/containers/Timestamp/index.test.jsx:97]: In addition to this can we have 2 more simple tests? Just ones that assert some timestamp in may outputs to something, and one in Jan. Then if we accidentally go to only BST or only UTC, one will fail

**Review by dr3 — CHANGES_REQUESTED** [review]

File: src/app/containers/Timestamp/helpers/testHelpers.js (hunk 3 above)
Comment (twinlensreflex) [reply src/app/containers/Timestamp/helpers/testHelpers.js]: I checked this and it seems [from the docs as if Moment's implementation of `isDST` is timezone/location aware](http://momentjs.com/docs/#/query/is-daylight-saving-time/). I could be misinterpreting that though.

Comment (dr3) [reply src/app/containers/Timestamp/helpers/testHelpers.js]: Looks like we need something like 
```
moment.tz(timestamp, 'Europe/London').isDST()
```

File: src/app/containers/Timestamp/index.test.jsx (hunk 4 above)
Comment (twinlensreflex) [reply src/app/containers/Timestamp/index.test.jsx]: good point!

File: src/app/containers/Timestamp/index.jsx [hunk 6]
```diff
@@ -1,20 +1,30 @@
...
 } from './timestampUtilities';
 import { GridItemConstrainedMedium } from '../../lib/styledGrid';
 
+const isToday = timestamp => {
+  const today = moment(Date.now());
+  return today.isSame(timestamp, 'day');
+};
+
+const formatType = timestamp =>
+  isToday(timestamp) ? alphaNumericDatetime : shortAlphaNumeric;
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/index.jsx]: Is it just me that struggles to read these names and know what the format will be?

We could rename:

- `alphaNumericDateTime` => `dateAndTime`
- `shortAlphaNumeric` => `dateOnly`

Naming is the hardest thing in computer science. These suggestions are just an opinion! What do you think?

File: src/app/containers/Timestamp/index.test.jsx [hunk 7]
```diff
@@ -38,46 +42,32 @@ describe('Timestamp', () => {
     />,
   );
 
-  it('should display only one timestamp when published === updated', () => {
+  it('should render only one timestamp when published === updated', () => {
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/index.test.jsx]:
```suggestion
  it('should render one date when published before today and never updated', () => {
```

I'd move this test below the next one too.

File: src/app/containers/Timestamp/index.test.jsx [hunk 8]
```diff
@@ -38,46 +42,32 @@ describe('Timestamp', () => {
...
     const renderedWrapper = renderedTimestamps(
       <Timestamp firstPublished={fifthJan} lastPublished={fifthJan} />,
     );
     expect(renderedWrapper[0].children[0].data).toEqual('5 January 2019');
     expect(renderedWrapper.length).toEqual(1);
   });
 
-  it('should display a relative timestamp when updated < 10 hours ago', () => {
+  it('should render one relative timestamp when published < 10 hours ago', () => {
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/index.test.jsx]:
```suggestion
  it('should render one relative timestamp when published < 10 hours ago and never updated', () => {
```

File: src/app/containers/Timestamp/index.test.jsx [hunk 9]
```diff
@@ -38,46 +42,32 @@ describe('Timestamp', () => {
...
-    expect(renderedWrapper.length).toEqual(2);
-    expect(renderedWrapper[0].children[0].data).toEqual('5 January 2019');
-    expect(renderedWrapper[1].children[0].data).toEqual('Updated 8 March 2019');
+    expect(renderedWrapper.length).toEqual(1);
+    expect(renderedWrapper[0].children[0].data).toEqual('6 hours ago');
   });
 
-  it('should render relative time if published < 10 hrs ago', () => {
-    const publishedFourHoursAgo = timestampGenerator({ hours: 4 });
+  it('should render one absolute timestamp (without datetime) when published yesterday or before', () => {
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/index.test.jsx]: This case is already covered by line 45. Delete?

File: src/app/containers/Timestamp/index.test.jsx [hunk 10]
```diff
@@ -95,13 +85,16 @@ describe('Timestamp', () => {
     expect(renderedWrapper[1].children[0].data).toEqual('Updated 4 hours ago');
   });
 
-  it('should render absolute time for lastPublished and for firstPublished if > 10 hrs ago', () => {
-    const firstPublishedTwelveHoursAgo = timestampGenerator({ hours: 12 });
-    const lastPublishedElevenHoursAgo = timestampGenerator({ hours: 11 });
+  it('should render absolute time (without datetime) for lastPublished and for firstPublished if published today > 10 hrs ago', () => {
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/index.test.jsx]:
```suggestion
  it('should render two dates (without time) when published and updated before today', () => {
```

File: src/app/containers/Timestamp/index.test.jsx [hunk 11]
```diff
@@ -111,4 +104,118 @@ describe('Timestamp', () => {
       /Updated [0-9]{1,2} \w+ [0-9]{4}/,
     );
   });
+
+  describe('time dependent tests', () => {
+    // eslint-disable-next-line no-unused-vars
+    let clock;
+    describe('tests after 10 am', () => {
+      const inBritishSummerTime = isBritishSummerTime(Date.now());
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/index.test.jsx:99]: Rather than using `lolex` to manipulate time, can we do this with Jest mocks?

Look at how I've done it here: https://github.com/bbc/simorgh/pull/1505/files#diff-aec09657c840fdae03a6bf73cec50c31R37

Should be able to do something like:

```js
const originalDateNow = Date.now;
Date.now = jest.fn(() => 1496235600000);

originalDateNow(); // if you need the _actual_ Date.now() value
Date.now(); // if you want the mocked value of 1496235600000
```

File: webpack.config.client.js [hunk 12]
```diff
@@ -109,6 +110,10 @@ module.exports = ({ resolvePath, IS_CI, IS_PROD, START_DEV_SERVER }) => {
           );
         },
       ),
+      new webpack.ContextReplacementPlugin(/moment[\\]locale$/, /en/),
```
Comment (ChrisBAshton) [inline webpack.config.client.js]: We need to keep our webpack config really, really well commented!

Can you add line-by-line inline comments explaining why all of this is needed please.

**Review by ChrisBAshton — CHANGES_REQUESTED** [review]
Good start 👏

File: src/app/containers/Timestamp/index.jsx (hunk 6 above)
Comment (twinlensreflex) [reply src/app/containers/Timestamp/index.jsx]: I agree 👍 I really didn't like the naming and I think those suggestions work.

File: src/app/containers/Timestamp/index.test.jsx [hunk 13]
```diff
@@ -91,24 +70,118 @@ describe('Timestamp', () => {
...
+          });
+          const renderedWrapper = renderedTimestamps(
+            <Timestamp
+              firstPublished={twentyThreeHoursAgo}
+              lastPublished={twentyThreeHoursAgo}
+            />,
+          );
+
+          expect(renderedWrapper.length).toEqual(1);
+          expect(renderedWrapper[0].children[0].data).toMatch(datetimeRegex);
```
Comment (dr3) [inline src/app/containers/Timestamp/index.test.jsx]:
```suggestion
          expect(renderedWrapper[0].children[0].data).toMatch(datetimeRegex);
          expect(renderedWrapper[0].children[0].data).toContain(descriptor);
```

File: src/app/containers/Timestamp/timestampUtilities.js [hunk 14]
```diff
@@ -12,31 +13,21 @@ export const isValidDateTime = dateTime => {
...
-
-export const formatUnixTimestamp = (milliseconds, formatType) => {
-  const dateObj = new Date(milliseconds);
-  const day = formatType.day(dateObj);
-  const month = formatType.month(dateObj);
-  const year = formatType.year(dateObj);
-  return formatType.format(day, month, year);
-};
+export const formatUnixTimestamp = (timestamp, momentString) =>
+  moment(timestamp)
```
Comment (dr3) [inline src/app/containers/Timestamp/timestampUtilities.js:18]: Can you add a basic "Expect moment to be called with X" test please

**Review by dr3 — CHANGES_REQUESTED** [review]

File: src/app/containers/Timestamp/timestampUtilities.js (hunk 14 above)
Comment (twinlensreflex) [reply src/app/containers/Timestamp/timestampUtilities.js:18]: you mean in `src/app/containers/Timestamp/timestampUtilities.test.js`?

File: src/app/containers/Timestamp/timestampUtilities.js [hunk 15]
```diff
@@ -12,31 +13,21 @@ export const isValidDateTime = dateTime => {
...
+    .tz('Europe/London')
+    .format(momentString);
 
 export const isTenHoursAgoOrLess = milliseconds => {
   const now = Date.now();
   return now - milliseconds <= 10 * 60 * 60 * 1000;
 };
+
+// 2019-03-22
+export const longNumeric = 'YYYY-MM-DD';
```
Comment (sareh) [inline src/app/containers/Timestamp/timestampUtilities.js]: It would be good to namespace these values. 
```suggestion
export const formatDateNumeric = 'YYYY-MM-DD';
```

File: src/app/containers/Timestamp/timestampUtilities.js [hunk 16]
```diff
@@ -12,31 +13,21 @@ export const isValidDateTime = dateTime => {
...
+
+// 22 March 2019
+export const dateOnly = 'D MMMM YYYY';
```
Comment (sareh) [inline src/app/containers/Timestamp/timestampUtilities.js]: Same here
```suggestion
export const formatDate = 'D MMMM YYYY';
```

File: src/app/containers/Timestamp/timestampUtilities.js [hunk 17]
```diff
@@ -12,31 +13,21 @@ export const isValidDateTime = dateTime => {
...
+
+// 22 March 2019, 17:05 BST
+export const dateAndTime = 'D MMMM YYYY, HH:mm z';
```
Comment (sareh) [inline src/app/containers/Timestamp/timestampUtilities.js]:
```suggestion
export const formatDateAndTime = 'D MMMM YYYY, HH:mm z';
```

File: src/app/containers/Timestamp/timestampUtilities.js [hunk 18]
```diff
@@ -12,31 +13,21 @@ export const isValidDateTime = dateTime => {
 
 export const leadingZero = val => (val < 10 ? `0${val}` : `${val}`);
```
Comment (sareh) [inline src/app/containers/Timestamp/timestampUtilities.js:15]: It would be useful to add a comment here explaining that when using this function, we recommend using webpack configuration to only load in the relevant timezone, rather than all of `moment-timezone` https://github.com/bbc/simorgh/blob/f22671bea4b78d3bf186723535af8ae948c801ff/webpack.config.client.js#L126-L128

**Review by sareh — CHANGES_REQUESTED** [review]
Great work with the test coverage. I've just added a few points regarding naming and adding comments. 

Screen Shot showing 100% test coverage

**Review by sareh — DISMISSED** [review]
Looks great! 👍

**Review by oluoluoxenfree — DISMISSED** [review]

**Review by dr3 — DISMISSED** [review]
Happy with the changes I requested, Not reviewed but happy with the other 2 approvals :)

**Comment by jamesdonoh** [comment]
Can we measure the effect on bundle size of this change? I know some analysis was done in https://github.com/bbc/simorgh/issues/637 but we should re-check now it has been implemented.

**Comment by twinlensreflex** [comment]
Bundle size investigation

*Bundle with no webpack plugins*
Screen Shot 2019-04-11 at 15 27 29
Screen Shot 2019-04-11 at 15 26 02

*With only `new MomentTimezoneDataPlugin({matchZones: 'Europe/London',})`*
Screen Shot 2019-04-11 at 15 25 33

*With only `new webpack.ContextReplacementPlugin(/moment[/\\]locale$/, /en-gb/)`*
Screen Shot 2019-04-11 at 15 24 36

*Post webpack plugin additions*:
Screen Shot 2019-04-11 at 15 28 10

Screen Shot 2019-04-11 at 15 28 19

**Review by oluoluoxenfree — APPROVED** [review]
Looks great :)

**Review by sareh — APPROVED** [review]
👍 Great work with the bundle size investigation.

**Comment by jamesdonoh** [comment]
Hi thanks for the changes, looks good. Useful as the visualisations are, do you have any figures to go with them? Would be interested to see a table of total bundle size for each, as context for our discussion about using moment.js more generally.

File: src/app/containers/README.md [hunk 19]
```diff
@@ -7,3 +7,26 @@ NB containers are just React components
...
+new webpack.ContextReplacementPlugin(/moment[/\\]locale$/, /en-gb/),
+/* moment-timezone-data-plugin allows you to specify how much
+* and what specific timezone data you wish to bundle.
+* matchZones: (string or array of strings) Only include data
+* for time zones with names matching this value.
+*/
+new MomentTimezoneDataPlugin({
+matchZones: 'Europe/London',
+}),
+```
```
Comment (jamesdonoh) [inline src/app/containers/README.md:32]: What's the thinking behind putting this info here? It doesn't seem directly related to this directory.

File: webpack.config.client.js [hunk 20]
```diff
@@ -109,6 +110,22 @@ module.exports = ({ resolvePath, IS_CI, IS_PROD, START_DEV_SERVER }) => {
...
         },
       ),
+      /*
+       * The webpack.ContextReplacementPlugin allows us to load only
+       * the specific locales we require into our code. By using this
+       * plugin we can reference `moment/locale/en.js` in the files
+       * we need it, without importing it in every file or importing
+       * every locale.
+       */
+      new webpack.ContextReplacementPlugin(/moment[/\\]locale$/, /en-gb/),
```
Comment (jamesdonoh) [inline webpack.config.client.js:120]: Did you consider using the `MomentLocalesPlugin` module that Moment.js recommend? Might be easier/safer to maintain than a regex solution.
https://momentjs.com/docs/#/use-it/webpack/

**Comment by jamesbrumpton** [comment]
LGTM

File: src/app/containers/README.md (hunk 19 above)
Comment (sareh) [reply src/app/containers/README.md:32]: @jamesdonoh We're going to pull this out into Psammead very soon, so it'll be recommended setup for any user of the psammead timestamp container.

File: webpack.config.client.js (hunk 20 above)
Comment (sareh) [reply webpack.config.client.js:120]: Looking at that documentation, I agree that it's a better solution. I'll write up a separate issue to refactor in a future PR.
